### PR TITLE
Refactor(streams): improve type safety and readability

### PR DESCRIPTION
**Pull request by k14uz**
### Changes:
- **Added JSDoc Comments**: Added JSDoc comments to interfaces and functions. This would help in understanding the purpose of interfaces and functions at a glance.
- **Variable Renaming for Clarity**: Renamed variables to be more descriptive.
  - Renamed `decoder` to `textDecoder`.
  - Renamed `parser` to `eventSourceParser`.
  - Renamed `fullResponse` to `aggregatedResponse`.
  - Renamed `res` to `response`.
- **Type Narrowing for Safety**: Implemented type narrowing to handle TypeScript errors related to property access on union types (e.g., `ParsedEvent | ReconnectInterval`). This ensures that we are accessing properties in a type-safe manner.
- **Created a Utility Function**: Extracted the logic for creating an empty `ReadableStream` into a separate utility function called `createEmptyReadableStream`. This adheres to the DRY principle.

Please let me know if there are any changes or adjustments you would like.

### CODE REVIEW

**Comment by sirlolcat** [comment]
I noticed that during the last PR, the prettier action failed as well; I would like to have more information on how I can stay consistent on prettier setup with the team. Thanks.

**Comment by sirlolcat** [comment]
I would also like to purpose the benefit of applying this refactor to the code-base as it grows to ensure it's maintainability. I would be happy to take that into action. Let me know about it.

**Review by jaredpalmer — CHANGES_REQUESTED** [review]
This is great! Can you add JSDoc comments to all exports? (but please leave further refactoring for another PR)

File: packages/core/streams/ai-stream.ts [hunk 1]
```diff
@@ -5,110 +5,142 @@ import {
   type ReconnectInterval
 } from 'eventsource-parser'
 
+/**
+ * Interface representing callbacks for an AIStream.
```
Comment (jaredpalmer) [inline packages/core/streams/ai-stream.ts:9]:
```suggestion
 * Helper callback methods for stream lifecycle events
```

File: packages/core/streams/ai-stream.ts [hunk 2]
```diff
@@ -5,110 +5,142 @@ import {
...
+ * @interface
+ */
 export interface AIStreamCallbacks {
   onStart?: () => Promise<void>
   onCompletion?: (completion: string) => Promise<void>
   onToken?: (token: string) => Promise<void>
 }
 
+/**
+ * Interface representing a custom parser for AIStream data.
```
Comment (jaredpalmer) [inline packages/core/streams/ai-stream.ts:19]: In comments, lets just say what is immediately and remove the `Interface representing...` parts

File: packages/core/streams/ai-stream.ts [hunk 3]
```diff
@@ -5,110 +5,142 @@ import {
...
 
     transform(chunk) {
-      parser.feed(decoder.decode(chunk))
+      eventSourceParser.feed(textDecoder.decode(chunk))
     }
   })
 }
 
 /**
- * This stream forks input stream, allowing us to use the result as a
```
Comment (jaredpalmer) [inline packages/core/streams/ai-stream.ts:47]: This is a more informative comment. Lets not lose important contributor information that could be helpful for newcomers.

File: packages/core/streams/ai-stream.ts [hunk 4]
```diff
@@ -5,110 +5,142 @@ import {
...
-export function trimStartOfStreamHelper() {
-  let start = true
-  return (text: string) => {
-    if (start) text = text.trimStart()
-    if (text) start = false
+/**
+ * Creates a function to trim the start of a stream.
+ * @return {Function} A function that trims the start of a stream.
+ */
+export function trimStartOfStreamHelper(): (text: string) => string {
```
Comment (jaredpalmer) [inline packages/core/streams/ai-stream.ts:98]: This removes valuable comment information. Let's not lose valuable information.

File: packages/core/streams/ai-stream.ts [hunk 5]
```diff
@@ -5,110 +5,142 @@ import {
...
+    if (isStreamStart) {
+      text = text.trimStart()
+      if (text) isStreamStart = false
+    }
     return text
   }
 }
 
+/**
+ * Creates an AIStream from the response.
```
Comment (jaredpalmer) [inline packages/core/streams/ai-stream.ts:111]: Comment needs more detail. Comments should not include the name of the method they are describing in the first sentence. Should be phrased like `Returns a ...`

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

**Comment by jaredpalmer** [comment]
On second thought, lets actually break this into different PRs. First lets fix type-safety and then move JSDoc comments to another PR

**Comment by sirlolcat** [comment]
@jaredpalmer if you are on the same page with the changes for type-safety, I can create a branch out of this one and move all the JSDocs there; let me know if that works out for you.

**Comment by jaredpalmer** [comment]
Need to run prettier

**Comment by jaredpalmer** [comment]
TS seems fine

**Comment by sirlolcat** [comment]
> Need to run prettier

All fine now.

**Comment by CarlosZiegler** [comment]
Great PR :)

**Review by MaxLeiter — APPROVED** [review]

**Comment by MaxLeiter** [comment]
Thank you!

**Pull request merged**